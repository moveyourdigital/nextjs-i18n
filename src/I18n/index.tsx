import useI18n, {
  I18nLocale,
  I18nParams,
  I18nPluralValue,
  I18nSingleValue,
} from "../hook/useI18n"
import { FC } from "react"

type I18nProps =
  | Record<I18nLocale, I18nSingleValue | I18nPluralValue> & {
      count?: number
      params?: I18nParams
    }

const I18n: FC<I18nProps> = (props) => {
  const { params, count, ...texts } = props
  const { t } = useI18n()
  if (texts.children) {
    throw new Error("I18n component must not have children")
  }
  return (
    <>
      {t(texts as Record<I18nLocale, I18nSingleValue | I18nPluralValue>, {
        params,
        count,
      })}
    </>
  )
}

export default I18n
