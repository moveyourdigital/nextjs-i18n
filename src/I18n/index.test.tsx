import { create } from "react-test-renderer"
import mockConsole from "jest-mock-console"
import I18n from "."

const i18n = {
  locale: "pt",
  locales: ["en", "pt"],
  defaultLocale: "en",
}

jest.mock("next/router", () => ({
  useRouter: () => i18n,
}))

describe("<I18n />", () => {
  beforeEach(() => {
    mockConsole()
  })

  describe("when passing pt and en", () => {
    it("renders pt text", () => {
      expect(create(<I18n pt="Olá" en="Hello" />).toJSON()).toMatchSnapshot()
    })
  })

  describe("when passing plurals", () => {
    it.each`
      count
      ${"0"}
      ${"1"}
      ${"2"}
      ${"3"}
      ${"42"}
    `("renders plural form of $count in the string", ({ count }) => {
      expect(
        create(
          <I18n
            pt={{
              0: "Zero artigos",
              1: "Um artigo",
              2: "Dois artigos",
              n: "Vários artigos",
            }}
            en={{
              0: "Zero articles",
              1: "One article",
              2: "Two articles",
              n: "Many articles",
            }}
            count={Number(count)}
          />,
        ).toJSON(),
      ).toMatchSnapshot()
    })
  })

  describe("when passing strings with placeholders", () => {
    it.each`
      count
      ${"0"}
      ${"1"}
      ${"2"}
      ${"42"}
    `("renders string with substitutes on $count messages", ({ count }) => {
      expect(
        create(
          <I18n
            pt="Olá, {{name}}. Tem {{count}} mensagens."
            en="Hi, {{name}}. You have {{count}} messages."
            params={{
              name: "John Doe",
              count,
            }}
          />,
        ).toJSON(),
      ).toMatchSnapshot()
    })
  })

  describe("when passing strings with placeholders", () => {
    it.each`
      count
      ${"0"}
      ${"1"}
      ${"2"}
      ${"42"}
    `("renders string with substitutes", ({ count }) => {
      expect(
        create(
          <I18n
            pt={{
              0: "Olá, {{name}}. Não tem mensagens.",
              1: "Olá, {{name}}. Tem {{count}} mensagem.",
              n: "Olá, {{name}}. Tem {{count}} mensagens.",
            }}
            en={{
              0: "Hi, {{name}}. You have no messages.",
              1: "Hi, {{name}}. You have {{count}} message.",
              n: "Hi, {{name}}. You have {{count}} messages.",
            }}
            params={{
              name: "John Doe",
              count,
            }}
            count={Number(count)}
          />,
        ).toJSON(),
      ).toMatchSnapshot()
    })
  })

  describe("when passing function and elements as placeholders", () => {
    it("renders string with right substitutes", () => {
      i18n.locale = undefined

      expect(
        create(
          <I18n
            en={
              'Hi, {{uppercase name.first}}. See your <link>messages</link> of <b>{{today "m/d/Y"}}</b>'
            }
            pt={
              'Olá {{name.first}}. Acesse suas <link>mensagens</link> do dia <b>{{today "d/m/Y"}}</b>'
            }
            params={{
              name: {
                first: "Jane",
                last: "Does",
              },
              link: <a href="/messages" />,
              b: <strong />,
              uppercase: (value: string) => {
                return value.toUpperCase()
              },
              today: (format: string) => {
                const date = new Date("2021-02-28")
                return format.replace(/d|m|Y/g, (match) => {
                  switch (match) {
                    case "d":
                      return date.getDate().toString()
                    case "m":
                      return (date.getMonth() + 1).toString()
                    case "Y":
                      return date.getFullYear().toString()
                  }
                })
              },
            }}
          />,
        ).toJSON(),
      ).toMatchSnapshot()

      expect(
        create(
          <I18n
            en="How did that happen? Maybe <0>this route</0> was valid in a point of time or the <1>matrix</1> had a glitch. Anyway, you can go back to our <2>homepage</2> to save the day."
            pt="Como é que isto aconteceu? É provável que <0>esta rota</0> tenha já existido algures no tempo ou então aconteceu um erro na <1>Matrix</1>. Seja como for, poderá voltar para a <2>homepage</2> e salvar o dia!"
            params={{
              0: <a href="/this-route" className="font-bold" />,
              1: <a href="/matrix" className="font-bold" />,
              2: <a href="/" className="font-bold" />,
            }}
          />,
        ).toJSON(),
      ).toMatchSnapshot()
    })
  })
})
