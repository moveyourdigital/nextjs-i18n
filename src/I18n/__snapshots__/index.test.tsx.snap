// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<I18n /> when passing function and elements as placeholders renders string with right substitutes 1`] = `
Array [
  "Hi, JANE. See your ",
  <a
    href="/messages"
  >
    messages
  </a>,
  "",
  " of ",
  <strong>
    2/28/2021
  </strong>,
  "",
  "",
]
`;

exports[`<I18n /> when passing function and elements as placeholders renders string with right substitutes 2`] = `
Array [
  "How did that happen? Maybe this route was valid in a point of time or the matrix had a glitch. Anyway, you can go back to our ",
  <a
    className="font-bold"
    href="/"
  >
    homepage
  </a>,
  "",
  " to save the day.",
]
`;

exports[`<I18n /> when passing plurals renders plural form of 0 in the string 1`] = `"Zero artigos"`;

exports[`<I18n /> when passing plurals renders plural form of 1 in the string 1`] = `"Um artigo"`;

exports[`<I18n /> when passing plurals renders plural form of 2 in the string 1`] = `"Dois artigos"`;

exports[`<I18n /> when passing plurals renders plural form of 3 in the string 1`] = `"Vários artigos"`;

exports[`<I18n /> when passing plurals renders plural form of 42 in the string 1`] = `"Vários artigos"`;

exports[`<I18n /> when passing pt and en renders pt text 1`] = `"Olá"`;

exports[`<I18n /> when passing strings with placeholders renders string with substitutes 1`] = `"Olá, John Doe. Não tem mensagens."`;

exports[`<I18n /> when passing strings with placeholders renders string with substitutes 2`] = `"Olá, John Doe. Tem 1 mensagem."`;

exports[`<I18n /> when passing strings with placeholders renders string with substitutes 3`] = `"Olá, John Doe. Tem 2 mensagens."`;

exports[`<I18n /> when passing strings with placeholders renders string with substitutes 4`] = `"Olá, John Doe. Tem 42 mensagens."`;

exports[`<I18n /> when passing strings with placeholders renders string with substitutes on 0 messages 1`] = `"Olá, John Doe. Tem 0 mensagens."`;

exports[`<I18n /> when passing strings with placeholders renders string with substitutes on 1 messages 1`] = `"Olá, John Doe. Tem 1 mensagens."`;

exports[`<I18n /> when passing strings with placeholders renders string with substitutes on 2 messages 1`] = `"Olá, John Doe. Tem 2 mensagens."`;

exports[`<I18n /> when passing strings with placeholders renders string with substitutes on 42 messages 1`] = `"Olá, John Doe. Tem 42 mensagens."`;
